receiver/googlecloudpubsubreceiver/receiver.go

```diff
@@ -87,11 +103,86 @@ func (receiver *pubsubReceiver) generateClientOptions() (copts []option.ClientOp
+
+	if receiver.config.Encoding != "" {
+		if receiver.consumerCount() > 1 {
+			return errors.New("cannot start receiver: multiple consumers were attached, but encoding was specified")
+		}
+		encodingID := convertEncoding(receiver.config.Encoding)
+		if encodingID == unknown {
+			extensionID := component.ID{}
+			err := extensionID.UnmarshalText([]byte(receiver.config.Encoding))
+			if err != nil {
+				return errors.New("cannot start receiver: neither a build in encoder, or an extension")
+			}
+			extensions := host.GetExtensions()
+			if extension, ok := extensions[extensionID]; ok {
+				if receiver.tracesConsumer != nil {
+					receiver.tracesUnmarshaler, ok = extension.(encoding.TracesUnmarshalerExtension)
+					if !ok {
+						return fmt.Errorf("cannot start receiver: extension %q is not a trace unmarshaler", extensionID)
+					}
+				}
+				if receiver.logsConsumer != nil {
+					receiver.logsUnmarshaler, ok = extension.(encoding.LogsUnmarshalerExtension)
+					if !ok {
+						return fmt.Errorf("cannot start receiver: extension %q is not a logs unmarshaler", extensionID)
+					}
+				}
+				if receiver.metricsConsumer != nil {
+					receiver.metricsUnmarshaler, ok = extension.(encoding.MetricsUnmarshalerExtension)
+					if !ok {
+						return fmt.Errorf("cannot start receiver: extension %q is not a metrics unmarshaler", extensionID)
+					}
+				}
+			} else {
+				return fmt.Errorf("cannot start receiver: extension %q not found", extensionID)
+			}
+		} else {
+			if receiver.tracesConsumer != nil {
+				switch encodingID {
+				case otlpProtoTrace:
+					receiver.tracesUnmarshaler = &ptrace.ProtoUnmarshaler{}
+				default:
+					return fmt.Errorf("cannot start receiver: build in encoding %s is not supported for traces", receiver.config.Encoding)
+				}
+			}
+			if receiver.logsConsumer != nil {
+				switch encodingID {
+				case otlpProtoLog:
+					receiver.logsUnmarshaler = &plog.ProtoUnmarshaler{}
+				case rawTextLog:
+					receiver.logsUnmarshaler = unmarshalLogStrings{}
+				case cloudLogging:
+					receiver.logsUnmarshaler = unmarshalCloudLoggingLogEntry{}
+				default:
+					return fmt.Errorf("cannot start receiver: build in encoding %s is not supported for logs", receiver.config.Encoding)
+				}
+			}
+			if receiver.metricsConsumer != nil {
+				switch encodingID {
+				case otlpProtoMetric:
+					receiver.metricsUnmarshaler = &pmetric.ProtoUnmarshaler{}
+				default:
+					return fmt.Errorf("cannot start receiver: build in encoding %s is not supported for metrics", receiver.config.Encoding)
+				}
+			}
+		}
+		createHandlerFn = receiver.createReceiverHandler
+	} else {
+		// we will rely on the attributes of the message to determine the signal, so we need all proto unmarshalers
+		receiver.tracesUnmarshaler = &ptrace.ProtoUnmarshaler{}
+		receiver.metricsUnmarshaler = &pmetric.ProtoUnmarshaler{}
+		receiver.logsUnmarshaler = &plog.ProtoUnmarshaler{}
+		createHandlerFn = receiver.createMultiplexingReceiverHandler
+	}
```

Comment:
There are a lot of branching conditions here, can you break it up into some helper functions.

Reply:
Pushed, update... but I have the impression there is something wrong with the CI/CD. Waiting until that is resolved.

Reply:
@TylerHelmuth I've split it into two functions (mainly grouping built-in and external encoders).